crypto/x509/x_attrib.c

```diff
@@ -56,3 +56,221 @@ X509_ATTRIBUTE *X509_ATTRIBUTE_create(int nid, int atrtype, void *value)
+        if (BIO_printf(out, "%*s", indent, "") <= 0) {
+            return 0;
+        }
+        return print_hex(out, av->value.octet_string->data,
+                 av->value.octet_string->length);
+
+    case V_ASN1_NULL:
+        return BIO_printf(out, "%*sNULL", indent, "");
+
+    case V_ASN1_OBJECT:
+        if (BIO_printf(out, "%*s", indent, "") <= 0) {
+            return 0;
+        }
+        return print_oid(out, av->value.object);
+    
+    /* ObjectDescriptor is an IMPLICIT GraphicString, but GeneralString is a
+    superset supported by OpenSSL, so we will use that anywhere a GraphicString
+    is needed here. */
+    case V_ASN1_GENERALSTRING:
+    case V_ASN1_GRAPHICSTRING:
+    case V_ASN1_OBJECT_DESCRIPTOR:
+        return BIO_printf(out, "%*s%.*s", indent, "",
+                          av->value.generalstring->length,
+                          av->value.generalstring->data);
+
+    /* EXTERNAL */
+    /* EMBEDDED PDV */
+
+    case V_ASN1_UTF8STRING:
+        return BIO_printf(out, "%*s%.*s", indent, "",
+                          av->value.utf8string->length,
+                          av->value.utf8string->data);
+
+    case V_ASN1_REAL:
+        return BIO_printf(out, "%*sREAL", indent, "");
+
+    /* RELATIVE-OID */
+    /* TIME */
+
+    case V_ASN1_SEQUENCE:
+        return ASN1_parse_dump(out, av->value.sequence->data,
+                        av->value.sequence->length, indent, 1);
+
+    case V_ASN1_SET:
+        return ASN1_parse_dump(out, av->value.set->data,
+                av->value.set->length, indent, 1);
+
+    /*
+        UTCTime ::= [UNIVERSAL 23] IMPLICIT VisibleString
+        GeneralizedTime ::= [UNIVERSAL 24] IMPLICIT VisibleString
+        VisibleString is a superset for NumericString, so it will work for that.
+    */
+    case V_ASN1_VISIBLESTRING:
+    case V_ASN1_UTCTIME:
+    case V_ASN1_GENERALIZEDTIME:
+    case V_ASN1_NUMERICSTRING:
+        return BIO_printf(out, "%*s%.*s", indent, "",
+                          av->value.visiblestring->length,
+                          av->value.visiblestring->data);
+
+    case V_ASN1_PRINTABLESTRING:
+        return BIO_printf(out, "%*s%.*s", indent, "",
+                          av->value.printablestring->length,
+                          av->value.printablestring->data);
+
+    case V_ASN1_T61STRING:
+        return BIO_printf(out, "%*s%.*s", indent, "",
+                          av->value.t61string->length,
+                          av->value.t61string->data);
+
+    case V_ASN1_IA5STRING:
+        return BIO_printf(out, "%*s%.*s", indent, "",
+                          av->value.ia5string->length,
+                          av->value.ia5string->data);
+
+    /* UniversalString */
+    /* CHARACTER STRING */
+
+    case V_ASN1_BMPSTRING:
+        value = OPENSSL_uni2asc(av->value.bmpstring->data,
+                                av->value.bmpstring->length);
+        int ret = BIO_printf(out, "%*s%s", indent, "", value);
```

Comment:
BMP String?  Isn't that potentially....large?  Nothing wrong with printing it if its properly converted to ascii of course, but is there value in printing what I am assuming to be a bitmap as a character string?

Reply:
This also appears to be an outdated view of my code. I removed this.